keps/sig-autoscaling/20190307-configurable-scale-velocity-for-hpa.md

```diff
@@ -0,0 +1,435 @@
+
+The main idea of the HPA is to autoscale because of a load increase to avoid request failures. It should work on both small clusters and large ones. For small clusters, the absolute number constraint works better (ScaleUpPods), for large clusters the percentage works better (ScaleUpPercentage).
+
+Example: If the current cluster size is `1` and calculated cluster size for this particular load is `20`, then we should reach it ASAP.
+
+For default values (ScaleUpPercent = 100, ScaleUpPods = 4) and “pick the largest constraint” concept, we’ll increase 1 -> 20 in 3 steps
+
+    1 -> 5 -> 10 -> 20
+
+The first step will use “scaleUp.pods” constraint; the next steps will use “scaleUp.percent” one.
+
+In case of more intuitive “pick the hardest limit” concept, we’ll increase the cluster in 6 steps:
+
+    1 -> 2 -> 4 -> 8 -> 12 -> 16 -> 20
+
+Given that each step takes [90 sec in worst case], we’ll respond to the load increase in `(6-3)*90 sec = ~ 5 min`.
+
+That’s too much, we should respond faster than that.
+
+[90 sec in worst case]: https://dzone.com/articles/kubernetes-autoscaling-101-cluster-autoscaler-hori-1
+
+The scale down constraints are a method to prevent too rapid loss of capacity. Hence, it makes sense to pick the maximum of two constraints.
+
+#### Stabilization Window
+
+Currently stabilization window ([PR][], [RFC][], [Algorithm Details][]) is used to gather “scale-down-recommendations” during some time (default is 5min),
+and a new number of replicas is set to the maximum of all recommendations.
+
+It may be defined by command line option `--horizontal-pod-autoscaler-downscale-stabilization-window`.
+
+The same idea is used in the `DelaySeconds` parameter above:
+
+- While scaling up, we should pick the safest (largest) "desiredReplicas" number during last `delaySeconds`.
+- While scaling down, we should pick the safest (largest) "desiredReplicas" number during last `delaySeconds`.
+
+The “stabilization window" as a result becomes an alias for the `constraints.scaleDown.delaySeconds`.
+
+[PR]: https://github.com/kubernetes/kubernetes/pull/68122
+[RFC]: https://docs.google.com/document/d/1IdG3sqgCEaRV3urPLA29IDudCufD89RYCohfBPNeWIM/edit#heading=h.3tdw2jxiu42f
+[Algorithm Details]: https://v1-14.docs.kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/#algorithm-details
+
+#### API Changes
+
+The following API changes are needed:
+
+We should add `Scale{Up,Down}{Percent,Pods}` fields into the HPA spec
+
+The resulting solution will look like this:
+
+```golang
+type HPAScaleConstraintValue struct {
+    Rate *HPAScaleConstraintRateValue
+    DelaySeconds *int32
+
+type HPAScaleConstraintRateValue struct {
+    Pods             *int32
+    Percent       *int32
+    PeriodSeconds * int32
+}
+
+type HPAScaleConstraints struct {
+       ScaleUp *HPAScaleConstraintValue
```

Comment:
Nit: extra spaces (8 instead of 4)

Reply:
Fixed: https://github.com/kubernetes/enhancements/pull/883/commits/0c6f599341ff28501d6156394d07760ed5baf777